v2/handler.js

```diff
@@ -0,0 +1,234 @@
+    self.completeOutOp = options.completeOutOp;
+    self.remoteHostPort = null; // filled in by identify message
+    self.lastSentFrameId = 0;
+}
+
+util.inherits(TChannelV2Handler, EventEmitter);
+
+TChannelV2Handler.prototype.nextFrameId = function nextFrameId() {
+    var self = this;
+    self.lastSentFrameId = (self.lastSentFrameId + 1) % v2.Frame.MaxId;
+    return self.lastSentFrameId;
+};
+
+TChannelV2Handler.prototype.handleFrame = function handleFrame(frame) {
+    var self = this;
+    switch (frame.body.type) {
+        case v2.Types.InitRequest:
+            return self.handleInitRequest(frame);
+        case v2.Types.InitResponse:
+            return self.handleInitResponse(frame);
+        case v2.Types.CallRequest:
+            return self.handleCallRequest(frame);
+        case v2.Types.CallResponse:
+            return self.handleCallResponse(frame);
+        case v2.Types.Error:
+            return self.handleError(frame);
+        default:
+            self.logger.error('unhandled frame type', {
+                type: frame.body.type
+            });
+    }
+};
+
+TChannelV2Handler.prototype.handleInitRequest = function handleInitRequest(reqFrame) {
+    var self = this;
+    if (self.remoteHostPort !== null) {
+        self.emit('error', new Error('duplicate init request')); // TODO typed error
+        return;
+    }
+    var hostPort = reqFrame.body.hostPort;
+    var processName = reqFrame.body.processName;
+    self.remoteHostPort = hostPort;
+    // TODO: use processName
+    self.emit('identify.in', hostPort, processName);
+    self.sendInitResponse(reqFrame);
+};
+
+TChannelV2Handler.prototype.handleInitResponse = function handleInitResponse(resFrame) {
+    var self = this;
+    if (self.remoteHostPort !== null) {
+        self.emit('error', new Error('duplicate init response')); // TODO typed error
+        return;
+    }
+    var hostPort = resFrame.body.hostPort;
+    var processName = resFrame.body.processName;
+    // TODO: use processName
+    self.remoteHostPort = hostPort;
+    self.emit('identify.out', hostPort, processName);
+};
+
+TChannelV2Handler.prototype.handleCallRequest = function handleCallRequest(reqFrame) {
+    var self = this;
+    var id = reqFrame.id;
+    var name = String(reqFrame.body.arg1);
+
+    if (self.remoteHostPort === null) {
+        self.emit('error', new Error('call request before init request')); // TODO typed error
+        return;
+    }
+
+    var handler = self.channel.getEndpointHandler(name);
+    var headers = {};
+
+    // TODO: couple reqFrame.body.tracing for all downstream requests
+    // TODO: implement service name?
+
+    self.runInOp(handler, {
+        id: id,
+        arg1: reqFrame.body.arg1,
+        arg2: reqFrame.body.arg2,
+        arg3: reqFrame.body.arg3,
+        headers: headers
+    }, function sendResponseFrame(err, res1, res2, callback) {
+        self.sendResponseFrame(reqFrame, headers, err, res1, res2, callback);
```

Comment:
why pass request headers to response frame ? Is this to enable threading of tracing ?